rust_src/src/lread.rs

```diff
@@ -309,4 +310,100 @@ pub fn eval_region(
     }
 }
 
+/// Read a character from the command input (keyboard or macro).
+/// It is returned as a number.
+/// If the character has modifiers, they are resolved and reflected to the
+/// character code if possible (e.g. C-SPC -> 0).
+///
+/// If the user generates an event which is not a character (i.e. a mouse
+/// click or function key event), `read-char' signals an error.  As an
+/// exception, switch-frame events are put off until non-character events
+/// can be read.
+/// If you want to read non-character events, or ignore them, call
+/// `read-event' or `read-char-exclusive' instead.
+///
+/// If the optional argument PROMPT is non-nil, display that as a prompt.
+/// If the optional argument INHERIT-INPUT-METHOD is non-nil and some
+/// input method is turned on in the current buffer, that input method
+/// is used for reading a character.
+/// If the optional argument SECONDS is non-nil, it should be a number
+/// specifying the maximum number of seconds to wait for input.  If no
+/// input arrives in that time, return nil.  SECONDS may be a
+/// floating-point value.
+#[lisp_fn(min = "0")]
+pub fn read_char(
+    prompt: LispObject,
+    inherit_input_method: LispObject,
+    seconds: LispObject,
+) -> LispObject {
+    if !prompt.is_nil() {
+        message_with_string!("%s", prompt, false);
+    }
+
+    let val =
+        unsafe { read_filtered_event(true, true, true, !inherit_input_method.is_nil(), seconds) };
+
```

Comment:
I believe something like this matches the original intent but is closer to idiomatic Rust if the return type is changed to `Option<EmacsInt>`. `read_char_exclusive` follows a similar pattern.

```
match val.into() {
    Some(num) => Some(unsafe { char_resolve_modifier_mask(num) }),
    None => None,
}
```